### PR TITLE
PoS integration

### LINKED ISSUE #474: Integrating PoS proposal into PNF3: proposer staking
Features:
- Add operations in the contracts for staking ETH. In the future decide if it's ETH or MATIC.
- Add endpoints for this proposers staking in optimist
- Add operations to sdk for this proposers staking
- This staking will be the base of the staking power when running spans of the weighted round robin algorithm for proposers selection in next task

Discuss:
- Tx fee payments could go to proposer's stake instead of adding it to proposer pending withdrawal.

### PR DESCRIPTION
**Pull request by daveroga**
Fixes #474 , #475, #488 

This PR integrates Proof of Stake into Nightfall following the proposal in #283 

## Features
- Add operations in the contracts for staking ETH. In the future decide if it's ETH or MATIC.
- Add endpoints for this proposers staking in optimist
- Add operations to sdk for this proposers staking
- This staking will be the base of the staking power when running spans of the weighted round robin algorithm for proposers selection in next task
- Implement weighted round robin algorithm based on the proposer staking for proposer selection

## Actions
- `registerProposer` moved to `stakeProposer(amount)`. This registers the proposer and if the proposer is already registered it increases the proposer stake amount.
- `deregisterProposer` moved to `unstakeProposer`. 
- No `REGISTRATION_BOND` now but a `MINIMUM_STAKE` (10 wei for testing now).
- `COOLING_OFF_PERIOD` moved to `CHALLENGE_PERIOD` (1 week).
- Changed messages from `require` statements in contracts to best practices used by Openzeppelin. `<contract name>: <message>`. So you know the contract where the message has been thrown. Useful now when debugging.
- `withdrawBond` moved to `withdrawStake` and used in the same way but with the stake of the proposer instead of the bond, so the proposer can do the withdraw of their stake. Previously to that, the proposer should call `unstakeProposer` to stop being proposer and wait this `CHALLENGE_PERIOD` to be allowed to do the withdraw.
- Moved `uint` to `uint256` in contracts to avoid errors and follow the same guide in all the contracts.
- `TimeLockedBond` moved to `TimeLockedStake` and added `challengeLocked` to store the amount locked by blocks proposed still in `CHALLENGE_PERIOD` and not claimed yet.
```
    struct TimeLockedStake {
        uint256 amount; // The amount held
        uint256 challengeLocked; // The amount locked by block proposed still in CHALLENGE_PERIOD
        uint256 time; // The time the funds were locked from
    }
```
- Need to request block payment for proposers. (#488).
- Add endpoing to optimist `/pending-payments` to get pending payments for a proposer.
- changeCurrentProposer will be based on this new proposer selection
- Define first approach with a set of parameter numbers for testnet (minimum stake, proposer count, number of springs in a span, ...) that we should analyze the suitable numbers for mainnet.
- Created test for testing PoS features. `npm run test-pos`

## Questions:
- Tx fee payments could go to proposer's stake instead of adding it to proposer pending withdrawal.

### CODE REVIEW

**Review by signorecello — CHANGES_REQUESTED** [review]

File: cli/src/proposer.mjs [hunk 1]
```diff
@@ -36,7 +36,7 @@ async function startProposer(testEnvironment) {
   await nf3.init(defaultMnemonic);
   if (await nf3.healthcheck('optimist')) console.log('Healthcheck passed');
   else throw new Error('Healthcheck failed');
-  await nf3.registerProposer();
+  await nf3.stakeProposer(1000);
```
[the remark this answers is not in the document]
Comment (daveroga) [reply cli/src/proposer.mjs:39]: Yes. I'll push now changes with a parameter for starting the proposer with `--stake <amount>`

File: nightfall-deployer/contracts/State.sol [hunk 2]
```diff
@@ -139,71 +153,220 @@ contract State is Structures, Config {
...
+      for (uint256 i = 0; i < weight; i++) {
+          slots.push(p.thisAddress);
+      }
+    }
+  }
+
+  /**
+   * @dev Shuffle the slots of all proposers
+   */
+  function shuffleSlots() internal {
```
[the remark this answers is not in the document]
Comment (daveroga) [reply nightfall-deployer/contracts/State.sol:332]: We can change randomness for shuffling slots with Chainlink VRF functionality. But this is a decision that will introduce the need of having LINK tokens and makes them available to fund requests to Chainlink VRF coordinators.

[the remark this answers is not in the document]
Comment (signorecello) [reply nightfall-deployer/contracts/State.sol:332]: I believe both would be good solutions for different stages of project maturity. In either case I wouldn't use a timestamp...

**Comment by ChaitanyaKonda** [comment]
Since we will be reviewing this when we are ready to have POS merged into master, can we close this PR which merges `daveroga/pos-proposal-integration2` into `feature-pos` and open a new PR when we want to merge `feature-pos` into `master` ? Until then we will have this as the branch `feature-pos`.

Otherwise, this would be sitting in pending PRs until ready for review which would be sometime until mainnet launch.

**Pull request merged**